[develop] Improve error messaging and logging for generate_FV3LAM_wflow.py

mkavulich opened the pull request:
This PR is now open for review

## DESCRIPTION OF CHANGES: 
The error messaging in the generate_FV3LAM_wflow.py has been overhauled to be more pythonic *and* more user-friendly. The old bash-like `print_err_msg_exit()` calls are replaced with proper exception handling, and `print_info_msg()` calls are replaced with the use of Python's built-in logging module. The latter has the added benefit of giving the same "tee"-like functionality--where messages are printed to screen and written to a log file simultaneously--without the need for messy subprocesses and the confusing error messages that come with them. 

Errors are now handled in a consistent way, with the errors clearly highlighted both on screen and in the log file, as well as an easy-to-follow exception traceback for those wanting that information. In addition, several error conditions that were previously **not** caught at the generate step now cause the script to fail with helpful error messages.

From the user perspective, functionality will not change, aside from differences when error conditions are present. Successful runs will remain the same, except that some log messaging has been clarified, and the log *file* contains additional information for each message, as well as some indentation changes for ease of reading.

### Example output
In this example, I specified an invalid variable "INVALID_VAR" in config.yaml. In the current code, the script *succeeds*, which is a dangerous condition that can produce unexpected results:
**Updated code**:
```
./generate_FV3LAM_wflow.py 

  ========================================================================
  Starting experiment generation...
  ========================================================================

  ========================================================================
  Starting function setup() in "setup.py"...
  ========================================================================

*********************************************************************
Experiment generation failed. See the error message(s) printed below.
For more detailed information, check the log file from the workflow
generation script: /mnt/lfs4/BMC/gsd-fv3-dev/kavulich/workdir/PR_414/ufs-srweather-app/ush/log.generate_FV3LAM_wflow
*********************************************************************

Traceback (most recent call last):
  File "./generate_FV3LAM_wflow.py", line 1037, in <module>
    generate_FV3LAM_wflow(USHdir, logfile)
  File "./generate_FV3LAM_wflow.py", line 103, in generate_FV3LAM_wflow
    setup()
  File "/mnt/lfs4/BMC/gsd-fv3-dev/kavulich/workdir/PR_414/ufs-srweather-app/ush/setup.py", line 107, in setup
    raise Exception(dedent(f'''
Exception: 
User-specified variable "INVALID_VAR" in config.yaml is not valid
Check config_defaults.yaml for allowed user-specified variables

```
For this same case, the log file shows the following text:
```
root                   DEBUG    Finished setting up debug file logging in /mnt/lfs4/BMC/gsd-fv3-dev/kavulich/workdir/PR_414/ufs-srweather-app/ush/log.generate_FV3LAM_wflow
root                   DEBUG    Logging set up successfully
generate_FV3LAM_wflow  INFO
  ========================================================================
  Starting experiment generation...
  ========================================================================
setup                  INFO
  ========================================================================
  Starting function setup() in "setup.py"...
  ========================================================================
root                   ERROR
*********************************************************************
Experiment generation failed. See the error message(s) printed below.
For more detailed information, check the log file from the workflow
generation script: /mnt/lfs4/BMC/gsd-fv3-dev/kavulich/workdir/PR_414/ufs-srweather-app/ush/log.generate_FV3LAM_wflow
*********************************************************************

Traceback (most recent call last):
  File "./generate_FV3LAM_wflow.py", line 1037, in <module>
    generate_FV3LAM_wflow(USHdir, logfile)
  File "./generate_FV3LAM_wflow.py", line 103, in generate_FV3LAM_wflow
    setup()
  File "/mnt/lfs4/BMC/gsd-fv3-dev/kavulich/workdir/PR_414/ufs-srweather-app/ush/setup.py", line 107, in setup
    raise Exception(dedent(f'''
Exception:
User-specified variable "INVALID_VAR" in config.yaml is not valid
Check config_defaults.yaml for allowed user-specified variables
```

In the log file you can see the above-mentioned additional information. Before each message, the subroutine writing the message is listed (a few top-level processes involving setting up logging and printing the error message are logged as "root"). Next on the line is a note of the "logging level"; see [Logging HOWTO](https://docs.python.org/3/howto/logging.html) for more details, but basically this denotes the "urgency" of the message, with "debug" being the lowest and "error" being the highest used here, indicating the script has failed in some way. While not implemented here, this will allow us to properly quarantine lower-priority printouts to the log file, or only print them to screen if DEBUG=TRUE, by making use of lower-priority levels such as logging.debug. And finally, log messages that extend to multiple lines are indented by two spaces, to make individual messages easier to parse visually.

### Type of change
- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update

## TESTS CONDUCTED: 
- [x] hera.intel
  - Attempted to run as many error condition checks as possible to ensure smooth error handling/messaging
    - [x] Ran ./generate_FV3LAM_wflow.py with a number of different failure conditions
      - [x] Prior to checking out code
      - [x] Prior to building code
        - Note: currently this does *not* exit with an error (in develop or with these new changes), as it would break existing tests. Will need to re-visit this in the future.
      - [x] Prior to loading regional_workflow conda environment
        - Note: On some platforms this error will be impossible to catch gracefully, due to not even being able to start running the script
      - [x] Removed Externals.cfg
      - [x] malformed Externals.cfg
      - [x] With no config.yaml
      - [x] With malformed config.yaml
      - [x] PREEXISTING_DIR_METHOD = quit
      - [ ] Incorrect config.yaml settings
        - [x] Did not include `MACHINE`
        - [x] Did not include `ACCOUNT` with `WORKFLOW_MANAGER=rocoto`
        - [x] Included invalid variable
        - [x] Mismatch between SPP array sizes
        - [x] Mismatch between LSM_SPP array sizes
        - [x] Incorrect PREDEF_GRID_NAME
        - [x] RESTART_INTERVAL set to non-int value
        - [x] DATE_FIRST_CYCL or DATE_LAST_CYCL set to non-date value
        - [x] USE_CUSTOM_POST_CONFIG_FILE true but CUSTOM_POST_CONFIG_FP does not exist
        - [x] USE_CRTM true but CRTM_DIR does not exist
        - [x] FCST_LEN_HRS > 999
        - [x] FCST_LEN_HRS not evenly divisible by LBC_SPEC_INTVL_HRS
        - [x] Mandatory variables (DT_ATMOS, LAYOUT_X, LAYOUT_Y, BLOCKSIZE) not set
        - [x] CCPP_PHYS_SUITE not valid
        - [x] EXTRN_MDL_SOURCE_BASEDIR_ICS, EXTRN_MDL_SOURCE_BASEDIR_LBCS set to non-existent location while USE_USER_STAGED_EXTRN_FILES is True
        - [x] RUN_TASK_VX_ENSGRID or RUN_TASK_VX_ENSPOINT set True while DO_ENSEMBLE is False
        - [x] With RUN_TASK_MAKE_GRID = False
          - [x] Set GRID_DIR to non-existent location
          - [x] Set GRID_DIR to ""
        - [x] With RUN_TASK_MAKE_OROG = False
          - [x] Set OROG_DIR to non-existent location
          - [x] Set OROG_DIR to ""
        - [x] With RUN_TASK_MAKE_SFC_CLIMO = False
          - [x] Set SFC_CLIMO_DIR to non-existent location
          - [x] Set SFC_CLIMO_DIR to ""
        - [x] Set PREEXISTING_DIR_METHOD to ""
      - [x] Incorrect machine file settings
        - [x] "MACHINE" does not correspond to a machine file
        - [x] machine file is malformed
    - [x] Ran ./run_WE2E_tests.sh with a number of different failure conditions
      - [x] Prior to checking out code
      - [x] Prior to loading regional_workflow conda environment
    - [x] Ran ALL WE2E tests, compared output to current develop
      - Develop: /scratch2/BMC/fv3lam/kavulich/UFS/workdir/issue_385/before_and_after_testing/develop/expt_dirs
      - This branch: /scratch2/BMC/fv3lam/kavulich/UFS/workdir/issue_385/before_and_after_testing/after/expt_dirs/
      - [x] No unexpected differences
- [ ] cheyenne.intel
    - [ ] Ran fundamental suite of WE2E tests
- [ ] cheyenne.gnu
    - [ ] Ran fundamental suite of WE2E tests
- [x] Jenkins testing
- [x] All CI checks passed

## DEPENDENCIES:
None

## DOCUMENTATION:
None, though the new, clearer error messaging and removal of extraneous comments could be considered a type of documentation change. 

Check the comments on #385 for more examples of the differences in error messages.

## ISSUE: 
Fixes #385 

## CHECKLIST
<!-- Add an X to check off a box. -->
- [ ] My code follows the style guidelines in the Contributor's Guide
- [ ] I have performed a self-review of my own code using the Code Reviewer's Guide
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] My changes need updates to the documentation. I have made corresponding changes to the documentation
- [ ] My changes do not require updates to the documentation (explain).
- [ ] My changes generate no new warnings
- [ ] New and existing tests pass with my changes
- [ ] Any dependent changes have been merged and published

## CONTRIBUTORS (optional): 
<!-- If others have contributed to this work aside from the PR author, list them here -->

mkavulich committed:
Initial logging cleanup in generate_FV3LAM_wflow.py

 - Introduce logging handler that prints messages to screen and log file
 - Remove unnecessary subprocess calls, everything can now be handled
pythonically through the logging module
 - Start removing calls to print_info_msg and print_err_msg_exit; we
want to raise and properly handle exceptions, not use sys.exit
 - Remove/condense outdated and/or overly verbose comments
 - Update unit tests for new calling structure

mkavulich committed:
More updates for better logging and error messages

 - Fix some functionality problems
   - Fix extraneous VERBOSE argument that I missed
   - NOMADS capability is clearly broken currently (calls the wrong
     filename), so for now move to a function that if called, returns
     an exception
   - Fix log file move
 - Re-enable running from other directories, did not realize this was
   necessary for WE2E tests
 - pass USHdir to generate_FV3LAM_wflow() and handle log file moving at
   the end of that function
 - Remove time info from log file to reduce clutter
 - Raise exception with trace if generate_FV3LAM_wflow() fails
   - This returns non-zero exit code, so external behavior is the same

mkavulich committed:
Set up logging in setup.py

 - Pass logger to setup()
 - Convert calls print_info_msg into logger.info
 - Convert calls to print_err_msg_exit to raise exceptions (specific
   where appropriate), preserving the error message
 - Condense some overly verbose comments
 - Do not print success message for individual scripts; could be
   confusing to users

mkavulich committed:
Add check that config.yaml contains only variables found in config_defaults.yaml

mkavulich committed:
Add checks for required python version and checks to run_WE2E_tests.sh

mkavulich committed:
More updates, fixes, and cleanup

 - Fix problem with valid config option check
 - Convert calls to print_err_msg_exit() in ush/python_utils/config_parser.py
 - Designate MACHINE as a "mandatory" variable; it *must* be specified
   in config.yaml
 - Formatting cleanup with dedent

mkavulich committed:
More error/logging cleanup in setup.py

 - Better error trap for config load failure
 - Explicitly raise error if machine file does not exist
 - Consolidate some more verbose comments
 - Print formatting improvements

mkavulich committed:
More error/logging cleanup in setup.py and other files

 - Change config_defaults value for ACCOUNT to empty string, so the
   logic to catch unset ACCOUNT will trigger properly
   - Also fix existing error message
 - Trap errors for incorrect PREDEF_GRID_NAME setting
 - Fix error messages for various SPP settings
 - Don't re-set USHdir
 - Consolidate and python-ize the check for missing machine file
   settings
 - Consolidate more verbose comments

danielabdi-noaa commented:
@mkavulich The unittest is failing probably due to a mixup of the community and nco test cases settings. I recall that when i first called `generate_workflow` directly,  it remembered the `RUN_ENVIR` setting from the first test and used it for the latter. So I was forced to run them in separate processes.

mkavulich committed:
More error/logging cleanup in setup.py

 - Consolidate several variable checks into loops over a list of
   variable names
 - Date variables are now a bit more flexible in format; change error
   message to match
 - Update and consolidate more comments

mkavulich committed:
Better logging name strategy, more error/logging cleanup in setup.py and other files

 - No longer pass new logging object to subroutines, simply allow
   functions to inherit logging and rename logging within the function
 - Eliminate print_info_msg and print_err_msg_exit from
   set_ozone_param.py
 - Remove unused and unneeded ALL_CDATES and NUM_CYCLES variables
 - Consolidate/eliminate more overly verbose comments
 - Change logic slightly for when RUN_TASK_MAKE_GRID,
   RUN_TASK_MAKE_OROG, or RUN_TASK_MAKE_SFC_CLIMO is false: First check
   if the fix dir is not specified and change to default location (with
   warning message), THEN check to see if the dir exists (and fail if it
   does not)

mkavulich committed:
Finish error/logging cleanup in setup.py, set_thompson_mp_fix_files.py

 - Consolidate more comments
 - Add warning for inline post turning off post-processing tasks
 - Consolidate unnecessary splitting of "settings" array

mkavulich committed:
Last set of initial changes! Now just need to test and review...

 - Remove/consolidate outdated or unnecessary comments
 - Remove print of rocoto details; it is now impossible for users to get
   this far *without* having loaded a module that includes rocoto
 - Add logging to get_crontab_contents.py

mkavulich committed:
First big change from more extensive testing:
Bug in check_var_valid_value.py caused an unrelated exception if the
intended check failed and err_msg was not specified. Turns out
that nowhere in the code is err_msg used, so let's get rid of it! In
addition, trying to check an empty string caused an *additional*
exception, so added some logic to handle that case.

mkavulich committed:
Add testing subprocess code back in per Daniel Abdi's advice; this is
necessary for CI testing

mkavulich committed:
Forgot to set up logger for check_for_preexist_dir_file.py; not sure why this didn't fail in manual tests...

mkavulich committed:
More fixes to Daniels tests

mkavulich commented:
@danielabdi-noaa I restored the multiprocessing to the unit tests in generate_FV3LAM_wflow.py, but I am still seeing errors that I don't understand (I can't figure out how to see the full logs). I am on PTO Monday and Tuesday but I may ask for a quick meeting after that to try to figure out what I did wrong.

danielabdi-noaa commented:
@mkavulich You can run the unittest locally 
```
python3 -m unittest generate_FV3LAM_wflow.py
```
It looks like you are just missing a `dedent` import in `get_crontab_contents.py`
```Process Process-2:
Traceback (most recent call last):
  File "/contrib/miniconda3/4.5.12/envs/regional_workflow/lib/python3.8/multiprocessing/process.py", line 313, in _bootstrap
    self.run()
  File "/contrib/miniconda3/4.5.12/envs/regional_workflow/lib/python3.8/multiprocessing/process.py", line 108, in run
    self._target(*self._args, **self._kwargs)
  File "/scratch2/BMC/gsd-hpcs/Daniel.Abdi/ufs-srweather-app/ush/generate_FV3LAM_wflow.py", line 494, in generate_FV3LAM_wflow
    add_crontab_line()
  File "/scratch2/BMC/gsd-hpcs/Daniel.Abdi/ufs-srweather-app/ush/get_crontab_contents.py", line 104, in add_crontab_line
    logger.info(dedent(
NameError: name 'dedent' is not defined
```

danielabdi-noaa commented:
By the way, is it possible to override how the logger prints messages so that it automatically dedents them just like `print_info_msg` does? That way we do not have to dedent every message going to the log file or the screen. There is a case where you have to dedent the message beforehand, e.g. when concatenating two strings, but in most cases it is convenient to just send the string as is.

mkavulich committed:
Add missing import to fix failing test

mkavulich commented:
@danielabdi-noaa Thanks for the info, the tests seem to be working now!

> By the way, is it possible to override how the logger prints messages so that it automatically dedents them just like `print_info_msg` does? That way we do not have to dedent every message going to the log file or the screen. There is a case where you have to dedent the message beforehand, e.g. when concatenating two strings, but in most cases it is convenient to just send the string as is.

There's no built-in way to do this as far as I can tell, but I could add a new function that does this. I didn't want to change the existing `print_info_msg()` or `print_err_msg_exit()` because I wanted to keep these initial improvements confined to the context of workflow generation and the existing functions are used throughout the system in several different contexts. I will put that in and then open this PR for review.

danielabdi-noaa commented:
@mkavulich Are you thinking of modifying the logger.info/debug behaviour to include dedentation automatically or a separate function? As you pointed out, there is no built-in way but there seem to be workarounds like [this](https://stackoverflow.com/a/66638269/1382302) that will automatically dedent both console and logfile output. I tried it quickly on your PR with this patch, and it looks like it does the job.

mkavulich committed:
Address some comments from Daniel: implement logging.info as its own
function with built-in dedenting.

mkavulich committed:
Fail with error if model executable does not exist

mkavulich committed:
Revert "Fail with error if model executable does not exist"

This reverts commit 252f91dda4d84bf4417d2af324282fd730625732.

mkavulich commented:
@danielabdi-noaa We could discuss a better strategy in the future but I don't want to "force" dedenting on all log messages, so I have implemented it as a separate function similar to `print_info_msg()` for this initial effort.

Thanks for all your comments here so far, I am opening this up for wider review but let me know if you have additional comments/suggestions.

venitahagerty commented:
Machine: hera
Compiler: intel
Job: WE
Repo location: /scratch1/BMC/zrtrr/rrfs_ci/autoci/pr/1085525584/20221020000517/ufs-srweather-app
Build was Successful
Rocoto jobs started
Long term tracking will be done on 9 experiments
If test failed, please make changes and add the following label back:
ci-hera-intel-WE
Experiment Failed on hera: grid_RRFS_CONUScompact_25km_ics_HRRR_lbcs_RAP_suite_RRFS_v1beta
2022-10-20 00:36:11 +0000 :: hfe05 :: Task make_orog, jobid=36829776, in state DEAD (FAILED), ran for 39.0 seconds, exit status=256, try=2 (of 2)
Experiment Failed on hera: grid_RRFS_CONUScompact_25km_ics_HRRR_lbcs_RAP_suite_HRRR
2022-10-20 00:36:13 +0000 :: hfe05 :: Task make_orog, jobid=36829774, in state DEAD (FAILED), ran for 51.0 seconds, exit status=256, try=2 (of 2)
Experiment Failed on hera: grid_RRFS_CONUS_25km_ics_FV3GFS_lbcs_RAP_suite_HRRR
2022-10-20 00:36:10 +0000 :: hfe11 :: Task make_orog, jobid=36829772, in state DEAD (FAILED), ran for 55.0 seconds, exit status=256, try=2 (of 2)
Experiment Failed on hera: grid_RRFS_CONUScompact_25km_ics_HRRR_lbcs_HRRR_suite_RRFS_v1beta
2022-10-20 00:36:06 +0000 :: hfe05 :: Task make_orog, jobid=36829779, in state DEAD (FAILED), ran for 38.0 seconds, exit status=256, try=2 (of 2)
Experiment Failed on hera: grid_RRFS_CONUS_25km_ics_FV3GFS_lbcs_FV3GFS_suite_GFS_v16
2022-10-20 00:36:14 +0000 :: hfe12 :: Task make_orog, jobid=36829780, in state DEAD (FAILED), ran for 41.0 seconds, exit status=256, try=2 (of 2)
Experiment Failed on hera: grid_RRFS_CONUScompact_25km_ics_HRRR_lbcs_HRRR_suite_HRRR
2022-10-20 00:36:10 +0000 :: hfe08 :: Task make_orog, jobid=36829770, in state DEAD (FAILED), ran for 53.0 seconds, exit status=256, try=2 (of 2)
Experiment Failed on hera: grid_RRFS_CONUS_25km_ics_GSMGFS_lbcs_GSMGFS_suite_GFS_v15p2
2022-10-20 00:36:12 +0000 :: hfe08 :: Task make_orog, jobid=36829778, in state DEAD (FAILED), ran for 40.0 seconds, exit status=256, try=2 (of 2)
Experiment Failed on hera: grid_RRFS_CONUS_25km_ics_FV3GFS_lbcs_FV3GFS_suite_GFS_v15p2
2022-10-20 00:36:09 +0000 :: hfe07 :: Task make_orog, jobid=36829775, in state DEAD (FAILED), ran for 43.0 seconds, exit status=256, try=2 (of 2)
Experiment Succeeded on hera: nco_grid_RRFS_CONUScompact_25km_ics_HRRR_lbcs_RAP_suite_HRRR
All experiments completed

mkavulich committed:
Restore "verbose" functionality now that we have a function

venitahagerty commented:
Machine: jet
Compiler: intel
Job: WE
Repo location: /lfs1/BMC/nrtrr/rrfs_ci/autoci/pr/1085525584/20221020000511/ufs-srweather-app
Build was Successful
Rocoto jobs started
Long term tracking will be done on 9 experiments
If test failed, please make changes and add the following label back:
ci-jet-intel-WE
Experiment Succeeded on jet: nco_grid_RRFS_CONUScompact_25km_ics_HRRR_lbcs_RAP_suite_HRRR
Experiment Succeeded on jet: grid_RRFS_CONUS_25km_ics_FV3GFS_lbcs_FV3GFS_suite_GFS_v15p2
Experiment Succeeded on jet: grid_RRFS_CONUScompact_25km_ics_HRRR_lbcs_RAP_suite_HRRR
Experiment Succeeded on jet: grid_RRFS_CONUS_25km_ics_FV3GFS_lbcs_RAP_suite_HRRR
Experiment Succeeded on jet: grid_RRFS_CONUScompact_25km_ics_HRRR_lbcs_RAP_suite_RRFS_v1beta
Experiment Succeeded on jet: grid_RRFS_CONUS_25km_ics_GSMGFS_lbcs_GSMGFS_suite_GFS_v15p2
Experiment Succeeded on jet: grid_RRFS_CONUS_25km_ics_FV3GFS_lbcs_FV3GFS_suite_GFS_v16
Experiment Succeeded on jet: grid_RRFS_CONUScompact_25km_ics_HRRR_lbcs_HRRR_suite_HRRR
Experiment Succeeded on jet: grid_RRFS_CONUScompact_25km_ics_HRRR_lbcs_HRRR_suite_RRFS_v1beta

mkavulich committed:
Missed one VERBOSE pass

mkavulich committed:
The f-string "=" specifier is not supported for python 3.6 (our minimum requirement); reverting to older format

venitahagerty commented:
Machine: jet
Compiler: intel
Job: WE
Repo location: /lfs1/BMC/nrtrr/rrfs_ci/autoci/pr/1085525584/20221020022018/ufs-srweather-app
Build was Successful
Rocoto jobs started
Long term tracking will be done on 9 experiments
If test failed, please make changes and add the following label back:
ci-jet-intel-WE
Experiment Succeeded on jet: nco_grid_RRFS_CONUScompact_25km_ics_HRRR_lbcs_RAP_suite_HRRR
Experiment Succeeded on jet: grid_RRFS_CONUS_25km_ics_FV3GFS_lbcs_RAP_suite_HRRR
Experiment Succeeded on jet: grid_RRFS_CONUScompact_25km_ics_HRRR_lbcs_RAP_suite_RRFS_v1beta
Experiment Succeeded on jet: grid_RRFS_CONUScompact_25km_ics_HRRR_lbcs_RAP_suite_HRRR
Experiment Succeeded on jet: grid_RRFS_CONUS_25km_ics_FV3GFS_lbcs_FV3GFS_suite_GFS_v15p2
Experiment Succeeded on jet: grid_RRFS_CONUS_25km_ics_GSMGFS_lbcs_GSMGFS_suite_GFS_v15p2
Experiment Succeeded on jet: grid_RRFS_CONUS_25km_ics_FV3GFS_lbcs_FV3GFS_suite_GFS_v16
Experiment Succeeded on jet: grid_RRFS_CONUScompact_25km_ics_HRRR_lbcs_HRRR_suite_HRRR
Experiment Succeeded on jet: grid_RRFS_CONUScompact_25km_ics_HRRR_lbcs_HRRR_suite_RRFS_v1beta

christinaholtNOAA commented on ush/generate_FV3LAM_wflow.py:
```diff
@@ -950,6 +916,11 @@ def generate_FV3LAM_wflow():
+
+    # Call script to get NOMADS data
+    if NOMADS:
+        get_nomads_data(NOMADS_file_type,EXPTDIR,USHdir,DATE_FIRST_CYCL,CYCL_HRS,FCST_LEN_HRS,LBC_SPEC_INTVL_HRS)
```
Consider putting each of these arguments on separate lines, also consider making them named arguments. This is quite a few arguments to manage as a single-line list.

christinaholtNOAA commented on ush/generate_FV3LAM_wflow.py:
```diff
@@ -950,6 +916,11 @@ def generate_FV3LAM_wflow():
 
         set_FV3nml_sfc_climo_filenames()
+
+    # Call script to get NOMADS data
```
script -> function

christinaholtNOAA commented on ush/generate_FV3LAM_wflow.py:
```diff
@@ -1043,146 +998,82 @@ def generate_FV3LAM_wflow():
+    mv_vrfy(logfile, EXPTDIR)
+
+def get_nomads_data(NOMADS_file_type,EXPTDIR,USHdir,DATE_FIRST_CYCL,CYCL_HRS,FCST_LEN_HRS,LBC_SPEC_INTVL_HRS):
+    print("Getting NOMADS online data")
```
A doc string would be useful here.

christinaholtNOAA commented on ush/generate_FV3LAM_wflow.py:
```diff
@@ -1043,146 +998,82 @@ def generate_FV3LAM_wflow():
+    NOMADS_script = os.path.join(USHdir, "NOMADS_get_extrn_mdl_files.sh")
+    # run_command(f"""{NOMADS_script} {date_to_str(DATE_FIRST_CYCL,format="%Y%m%d")} \
+    #                 {date_to_str(DATE_FIRST_CYCL,format="%H")} {NOMADS_file_type} {FCST_LEN_HRS} {LBC_SPEC_INTVL_HRS}""")
+    raise Exception("Nomads script does not work")
```
Hahahaha! Maybe giving some feedback earlier than this would be useful!

christinaholtNOAA commented on ush/generate_FV3LAM_wflow.py:
```diff
@@ -1043,146 +998,82 @@ def generate_FV3LAM_wflow():
+                        filemode='w')
+    logging.debug(f'Finished setting up debug file logging in {logfile}')
+    console = logging.StreamHandler()
+    console.setLevel(logging.INFO)
```
Does this mean we won't see errors on the screen?

mkavulich replied:
No, all messages of "INFO" level or higher are printed to screen; this includes ERROR and WARNING messages as well.  https://docs.python.org/3/library/logging.html#levels

I hope to later do a sanitation of the workflow to demote some messages to "DEBUG" level, and have those only print to the log file unless the user specifies otherwise. That is for future work though.

christinaholtNOAA replied:
Thanks for the clarification!

christinaholtNOAA commented on ush/generate_FV3LAM_wflow.py:
```diff
@@ -1043,146 +998,82 @@ def generate_FV3LAM_wflow():
+        generate_FV3LAM_wflow(USHdir, logfile)
+    except:
+        # If the call to the generate_FV3LAM_wflow function above was not successful, 
+        # print out an error message and exit with a nonzero return code.
```
Is this comment necessary? Seems duplicative and obvious given the code.

christinaholtNOAA commented on ush/python_utils/print_msg.py:
```diff
@@ -39,3 +40,25 @@ def print_info_msg(info_msg, verbose=True):
         return True
     return False
+
+def log_info(info_msg, verbose=True, ddent=True):
```
You might consider `dedent_` as this variable name. It's a fairly common approach to dealing with namespace issues that doesn't make you wonder if it was misspelled, and is clear that you are avoiding a namespace collision.

christinaholtNOAA commented on ush/python_utils/print_msg.py:
```diff
@@ -39,3 +40,25 @@ def print_info_msg(info_msg, verbose=True):
+        None
+    """
+
+    # "sys._getframe().f_back.f_code.co_name" returns the name of the calling function
```
Consider removing commented lines of code.

mkavulich replied:
This isn't code, it's a comment clarifying the use of obscure python functionality

christinaholtNOAA replied:
Oops!

christinaholtNOAA commented on ush/set_ozone_param.py:
```diff
@@ -92,12 +90,11 @@ def set_ozone_param(ccpp_phys_suite_fp):
         ozone_param = "ozphys"
     else:
-        print_err_msg_exit(
+        raise Exception(
```
Does KeyError make sense here?

christinaholtNOAA commented on ush/setup.py:
```diff
@@ -84,30 +85,38 @@ def setup():
+
+    # Load the user config file, containing contains user-specified values 
+    # variables that override their default values. Ensure all user-specified
+    # variables correspond to a default value. 
```
This comment could use some work.

christinaholtNOAA commented on ush/setup.py:
```diff
@@ -84,30 +85,38 @@ def setup():
+                        Could not load YAML config file:  {EXPT_CONFIG_FN}
+                        The file may be formatted incorrectly; reference the Users Guide for more info 
+                        ''')
+        raise Exception(errmsg) from None
```
This is going to cause some problems for debugging if you can't see the original exception raised when loading the config file fails. It would be useful to catch specific failures here and raise them. 

Perhaps the file is formatted correctly but the user doesn't have appropriate permissions? Or a whole host of other things?

mkavulich replied:
I have returned the stack trace for this particular error. It is very ugly however, and I would like to find a cleaner solution in the future

christinaholtNOAA commented on ush/setup.py:
```diff
@@ -84,30 +85,38 @@ def setup():
+    for key in cfg_u:
+        if key not in flatten_dict(cfg_d):
+            raise Exception(dedent(f'''
+                            User-specified variable "{key}" in {EXPT_CONFIG_FN} is not valid
```
Suggest adding a period at the end of the line.

christinaholtNOAA commented on ush/setup.py:
```diff
@@ -84,30 +85,38 @@ def setup():
+
+    import_vars(dictionary=cfg_u, env_vars=["MACHINE",
+                   "EXTRN_MDL_NAME_ICS", "EXTRN_MDL_NAME_LBCS",
+                   "FV3GFS_FILE_FMT_ICS", "FV3GFS_FILE_FMT_LBCS"])
```
This list would be helpful with one item per line.

danielabdi-noaa replied:
This is my fault. I often let the linter/formatter take care of it and it would be good to have that or something similar as a precommit script.

christinaholtNOAA commented on ush/setup.py:
```diff
@@ -402,26 +425,17 @@ def get_location(xcs,fmt):
-            You can specify the missing location(s) in config.sh"""
-        )
+    # Mandatory variables *must* be set in the user's config or the machine file; the default value is invalid
+    mandatory = ['NCORES_PER_NODE', 'FIXgsm', 'FIXaer', 'FIXlut', 'TOPO_DIR', 'SFC_CLIMO_INPUT_DIR']
```
This is a long enough list that one item per line is probably easier to manage.

mkavulich replied:
That seems like a way to blow up a little code into a lot of lines. Is this recommended by a standard or just personal preference?

christinaholtNOAA replied:
It's not just a personal preference. It's easier to maintain the list over time as we can clearly see when the list is changed in a variety of different `diff` tools. I'm cool with you not making the change here, but I will likely address it in an upcoming PR.

christinaholtNOAA commented on ush/setup.py:
```diff
@@ -402,26 +425,17 @@ def get_location(xcs,fmt):
+    mandatory = ['NCORES_PER_NODE', 'FIXgsm', 'FIXaer', 'FIXlut', 'TOPO_DIR', 'SFC_CLIMO_INPUT_DIR']
+    for val in mandatory:
+        # globals() returns dictionary of global variables
+        if not globals()[val]:
```
It is more efficient to call globals once and assign it to a variable outside this loop.

christinaholtNOAA commented on ush/setup.py:
```diff
@@ -505,123 +512,62 @@ def get_location(xcs,fmt):
-    # Make sure RESTART_INTERVAL is set to an integer value if present
-    #
-    # -----------------------------------------------------------------------
-    #
```
Any chance you'd be willing to make this sort of change throughout this file? The 7 lines to 1 line is clearer.

mkavulich replied:
I feel that's outside the scope of this PR, the whole workflow needs a comment trim

christinaholtNOAA replied:
That's why I asked.

christinaholtNOAA commented on ush/setup.py:
```diff
@@ -505,123 +512,62 @@ def get_location(xcs,fmt):
-            print_err_msg_exit(
+        try:
+            if not os.path.exists(CUSTOM_POST_CONFIG_FP):
+                raise
```
What are you raising here? Why does it need to be in a try block?

mkavulich replied:
Because I'd like to check for the existence of this file and error out appropriately here in this script, even though we don't need to take any action on this file until later in the workflow. Do you know of a more elegant way to do this?

christinaholtNOAA replied:
My first thought is to raise the exception inside the if block if the file doesn't exist. I know you'd mentioned the logger.exception needs to be used in a try block though. You don't seem to be using that here. Just curious.

mkavulich replied:
Ah, I see the issue. os.path.exists returns an exception if passed "None", so this structure allows us to catch all errors in one. I added a clarifying comment for future developers.

christinaholtNOAA reviewed (commented):
Nice!

danielabdi-noaa commented on tests/WE2E/run_WE2E_tests.sh:
```diff
@@ -65,6 +65,63 @@ WE2Edir="$TESTSdir/WE2E"
+fi
+#
+#-----------------------------------------------------------------------
+#
```
`generate_workflow.py` does these checks so I wonder if we can somehow do without this piece of code to avoid duplication? However, I do understand the idea is to catch errors as early as possible! It just looks odd to me that we are checking python packages from bash, so maybe putting them in a separate python script and calling that will work?

mkavulich replied:
@danielabdi-noaa This script calls get_WE2Etest_names_subdirs_descs.sh which calls $USHdir/calculate_cost.py, which at the very least needs pyyaml (and maybe more, I haven't checked further than that). So we need python checks at the beginning of this script to check for a graceful exit.

Hopefully soon this will be python-based end-to-end so these checks can be neater.

danielabdi-noaa commented on ush/generate_FV3LAM_wflow.py:
```diff
@@ -459,7 +445,7 @@ def generate_FV3LAM_wflow():
             )
         except:
-            print_err_msg_exit(
+            logging.exception(
```
There is this NCO requirement that "ERROR:" be printed so just want to make sure the exception mechanism do that. The error messages look good to me but you may want to check with NCO folks. @MatthewPyle-NOAA

MatthewPyle-NOAA replied:
@danielabdi-noaa  From the implementation standards document:

> The context of the error must be communicated as descriptively as possible and prefaced with “WARNING:” or “FATAL ERROR:”


But honestly for this workflow generating functionality that won't run routinely in operations, it probably isn't a huge concern

mkavulich replied:
I think this is a good idea actually, even if it is not technically needed. I will add a "FATAL ERROR:" line to the beginning of the catch-all error message, for extra "oomph" to catch the user's attention.

danielabdi-noaa commented on ush/generate_FV3LAM_wflow.py:
```diff
@@ -1043,146 +998,82 @@ def generate_FV3LAM_wflow():
-    # put to the "tee" command).  Then remove the temporary file.
-    #
-    (_, exptdir, _) = run_command(f'''sed "1q;d" "{tmp_fp}"''')
-    (_, retval, _) = run_command(f''' sed "2q;d" "{tmp_fp}"''')
```
I am glad this most un-pythonic piece of code is gone! There are still a couple of `run_commands` left in the crontab management code, which hopefully can be replaced with a python crontab management package instead.

danielabdi-noaa commented on ush/python_utils/print_msg.py:
```diff
@@ -39,3 +40,25 @@ def print_info_msg(info_msg, verbose=True):
+            logger.info(indent(dedent(info_msg), '  '))
+        else:
+            logger.info(info_msg)
+
```
Will there be a log_debug counterpart for logging.debug?

mkavulich replied:
Yes, I plan on introducing "debug" functionality later; right now it's in a weird "in-between" where VERBOSE and DEBUG are specified in the config file, but I'd like to do away with that and make it a command-line argument. But that will require a lot more discussion so I didn't want to make that part of this PR (especially since it's a "minor release" that shouldn't make major changes to how the workflow is set up and run). Now that you mention it, I will open up an issue for that so it gets tracked.

danielabdi-noaa reviewed (commented):
@mkavulich Thanks for your great work! This PR is a nice demonstration of the power of python workflow compared to shell workflow. I have left a few minor suggestions and will approve once I am done with testing the error reporting cases you listed.

mkavulich committed:
Address reviewer comments

mkavulich commented:
@christinaholtNOAA  I believe I have addressed most of your concerns and followed up on the rest with a few comments/questions. Let me know what you think.

venitahagerty commented:
Machine: jet
Compiler: intel
Job: WE
Repo location: /lfs1/BMC/nrtrr/rrfs_ci/autoci/pr/1085525584/20221021213518/ufs-srweather-app
If test failed, please make changes and add the following label back:
ci-jet-intel-WE

christinaholtNOAA commented on ush/setup.py:
```diff
@@ -400,26 +427,18 @@ def get_location(xcs,fmt):
+    mandatory = ['NCORES_PER_NODE', 'FIXgsm', 'FIXaer', 'FIXlut', 'TOPO_DIR', 'SFC_CLIMO_INPUT_DIR']
+    globalvars = globals()
+    for val in mandatory:
+        # globals() returns dictionary of global variables
```
Maybe move this comment up and out of the loop with the call to `globals()`?

christinaholtNOAA reviewed (approved):
Everything looks good to me. The last two items (a comment below and an unresolved previous comment) are purely cosmetic.

venitahagerty commented:
Machine: hera
Compiler: intel
Job: WE
Repo location: /scratch1/BMC/zrtrr/rrfs_ci/autoci/pr/1085525584/20221021213515/ufs-srweather-app
Build was Successful
Rocoto jobs started
Long term tracking will be done on 9 experiments
If test failed, please make changes and add the following label back:
ci-hera-intel-WE
Experiment Succeeded on hera: grid_RRFS_CONUScompact_25km_ics_HRRR_lbcs_RAP_suite_RRFS_v1beta
Experiment Succeeded on hera: grid_RRFS_CONUS_25km_ics_FV3GFS_lbcs_FV3GFS_suite_GFS_v15p2
Experiment Succeeded on hera: grid_RRFS_CONUS_25km_ics_GSMGFS_lbcs_GSMGFS_suite_GFS_v15p2
Experiment Succeeded on hera: nco_grid_RRFS_CONUScompact_25km_ics_HRRR_lbcs_RAP_suite_HRRR
Experiment Succeeded on hera: grid_RRFS_CONUScompact_25km_ics_HRRR_lbcs_RAP_suite_HRRR

mkavulich committed:
Comment to clarify use of manual "raise" in setup.py

mkavulich committed:
One more comment addressed

mkavulich commented:
@danielabdi-noaa I think you still have one un-resolved comment as well, let me know if everything looks good!

danielabdi-noaa commented:
I've now run almost all error testing cases listed in the PR description and they work as expected so approving. 

Only the following case looks like it is impossible to catch on Hera atleast. If we run a WE2E test case though without conda activation, it would have been probably caught!
- Prior to loading regional_workflow conda environment

This one is disabled to get unittests working and i think for now it should be that way in my opinion. I've had to disable a couple of hard exits to get unittests working before, but if the benefits outweight the cons we could get rid of the unittests.
- Prior to building code

mkavulich commented:
@danielabdi-noaa Agreed on both counts. It would be nice to implement a check that the code is built, but I agree we will need to think of a better way to sync that with automated tests.
